development/build/utils.js

```diff
@@ -287,6 +287,21 @@ function getBuildIcon({ buildType }) {
   const svg = readFileSync(svgLogoPath, 'utf8');
   return `data:image/svg+xml,${encodeURIComponent(svg)}`;
 }
+/**
+ * Takes the given JavaScript file at `filePath` and replaces its contents with
+ * a script that injects the original file contents into the document in which
+ * the file is loaded. Useful for extensions to run scripts synchronously in the
+ * "MAIN" world.
+ *
+ * @param {string} filePath - The path to the file to convert to a self-injecting
+ * script.
+ */
+function makeSelfInjecting(filePath) {
+  const fileContents = readFileSync(filePath, 'utf8');
+  const textContent = JSON.stringify(fileContents);
+  const html = `{let d=document,s=d.createElement('script');s.textContent=${textContent};d.documentElement.appendChild(s).remove();}`;
+  writeFileSync(filePath, html, 'utf8');
```

Comment:
```suggestion
  const js = `{let d=document,s=d.createElement('script');s.textContent=${textContent};d.documentElement.appendChild(s).remove();}`;
  writeFileSync(filePath, js, 'utf8');
```

Reply:
Since we left `src` adventures (document-write, sync-xhr, etc) and circled back to `textContent`, I'd like to resurface my original suggestion of how to address this:

```javascript
function makeSelfInjecting(filePath) {
  const fileContents = readFileSync(filePath, 'utf8');
  const html = `{let d=document,s=d.createElement('script');s.textContent='init();' + init.toString();d.documentElement.appendChild(s).remove();function init(){${fileContents}}}`;
  writeFileSync(filePath, html, 'utf8');
}
```

I prefer this because it doesn't require stringifying the JS content, nor injecting it into a specific JS operation which leaves it cleaner and clearer.

Reply:
`inpage.js` has already been minified at this stage; I don't see a reason to make the output file cleaner and clearer.

Wrapping it this way pollutes the global namespace with an `init` function; so we'd need to further wrap this in an IIFE. So it'd now be wrapped 3 times.

I don't see any reason to avoid stringifying the text content during compilation. Why would this be undesirable?

With your proposed change the `inpage.js` JS will need to be parsed twice by the user's browser: once in the wrapped `init` function, and once again after being injected into the script's `textContent`. I don't see any reason for this.


Let me know if there is something here I'm missing.

Reply:
You are right about everything you say.
Everything I argue is also correct.
All upsides and downsides of both of our suggestions are quite minor.
At this point it's nothing more than a matter of different approaches, your approach is definitely just as fine to go with :)